src/main/kotlin/uk/gov/dluhc/printapi/service/PrintResponseFileService.kt

```diff
@@ -17,11 +19,28 @@ class PrintResponseFileService(
         val printResponsesString = sftpService.fetchFileFromOutBoundDirectory(directory, fileName)
         val printResponses = parsePrintResponseContent(printResponsesString)
         printResponseProcessingService.processBatchAndPrintResponses(printResponses)
-        sftpService.removeFileFromOutBoundDirectory(directory, fileName)
+        removeFile(directory, fileName)
     }
 
     private fun parsePrintResponseContent(printResponsesString: String): PrintResponses {
         logger.debug { "Parsing print responses $printResponsesString" }
         return objectMapper.readValue(printResponsesString, PrintResponses::class.java)
     }
+
+    private fun removeFile(directory: String, fileName: String) {
+        try {
+            sftpService.removeFileFromOutBoundDirectory(directory, fileName)
+        } catch (ex: MessagingException) {
```

Comment:
TL;DR;
Change the method in SftpService to declare it throws `IOException`, catch `IOException` here rather than `MessagingException`, log a `warn` rather than `error`, don't re-throw the exception back.

---

Vishal had a similar challenge re: exception handling from SFTP with a ticket he was working on in this project recently; and we agreed (I convinced him 😁 ) that the fact that the Spring SFTP Template class used in `SftpService` throws a "messaging" related exception is just horrible!
It feels wrong that our service classes need to know about Spring messaging.

Have a look in `SftpService` at what Vishal did, but basically he did this:
```
    /**
     * Renames the file by suffixing ".processing" to its original name. Returns the newly renamed file name
     * Note: The "/" is appended between fileDirectoryPath and fileName before it's renamed
     * @param directory the location of the status file to be renamed
     * @param originalFileName the name of the file e.g. fileName.json
     * @throws IOException while renaming the file.
     */
    @Throws(IOException::class)
    fun markFileForProcessing(directory: String, originalFileName: String): String {
```
Therefore the client class (the thing that is calling the `SftpService` method) now only needs to catch and process `IOException`
That would be good if you could do a similar thing here.

The next thing is what to do if SftpService throws an exception that is not caused by file not found (ie. your rethrow on line 37). I'm not massively keen on that as we'll have an exception being thrown back up the stack - are we catching it and processing it anywhere up the stack?

I guess we need to consider what the context is. This is being called from `processPrintResponseFile` which is about processing a single print response file (which contains a collection of Batch Responses and a collection of Print Responses); and we are trying to remove the file as the very last step. By this point we have already processed all of the Batch Responses, and processed all of the Print Responses by putting them on a queue.

If the exception is because the file is not found (other than it being very weird because we had already read and processed it .... but I do get that it _could_ happen) ... it doesn't feel like an `error`. There's nothing we can do about it, and it actually doesn't matter because we have processed the file. It's just that something else has removed it from under us. Feels like a `warn` to me?

Hmmmm, now, in the case of it not being based on file not found ... thats something different altogether ....

Rethrowing it is probably not the thing to do though. Because `printResponseFileService.processPrintResponseFile` is called from a listener, if we rethrow the exception, the listener will throw it back, the message will not get ACK'ed, and it will get re-queued or DLQ'ed if it's failed 5 times. Neither of which feel right?
As mentioned above, by this point - irrespective of what the exception type/cause is - we have already successfully processed the file and it's contents.

I think we'd be better off by not re-throwing, and logging something in the `else` block. _Maybe_ an `error` ?, but again, there's nothing we can do about it. It's probably a `warn` ?

Does that make sense?

Reply:
Thanks, @NathanRussellValtech , for the detailed suggestion. 

I will change the `SftpService` `removeFileFromOutBoundDirectory` method signature to throw `IOException` and will change the logging for file not found to a warning and for the other IO errors don't throw the exception, but just log it as an error.